### PR TITLE
[Merged by Bors] - feat: port Data.Matrix.PEquiv

### PR DESCRIPTION
**Pull request by Ruben-VandeVelde**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Data/Matrix/PEquiv.lean [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+def toMatrix [DecidableEq n] [Zero α] [One α] (f : m ≃. n) : Matrix m n α :=
+  of fun i j => if j ∈ f i then (1 : α) else 0
+#align pequiv.to_matrix PEquiv.toMatrix
+
+-- TODO: set as an equation lemma for `toMatrix`, see mathlib4#3024
+@[simp]
+theorem toMatrix_apply [DecidableEq n] [Zero α] [One α] (f : m ≃. n) (i j) :
+    toMatrix f i j = if j ∈ f i then (1 : α) else 0 :=
+  rfl
+#align pequiv.to_matrix_apply PEquiv.toMatrix_apply
```
Comment (semorrison) [inline Mathlib/Data/Matrix/PEquiv.lean:69]: Is this something we need to do before merging?

Comment (Ruben-VandeVelde) [reply Mathlib/Data/Matrix/PEquiv.lean:69]: I don't think we can do this before #3024 lands. CC @eric-wieser

Comment (eric-wieser) [reply Mathlib/Data/Matrix/PEquiv.lean:69]: Yes, we don't need to do this before merging. As long as we leave behind the references to 3024, then there is no rush to actually merge the PR. We can do a second pass over these comments later.

File: Mathlib/Data/Matrix/PEquiv.lean [hunk 2]
```diff
@@ -50,13 +50,18 @@ variable {α : Type v}
 
 open Matrix
 
-/-- `to_matrix` returns a matrix containing ones and zeros. `f.to_matrix i j` is `1` if
+-- Porting note: added
+instance [DecidableEq n] (j : n) (o : Option n) : Decidable (j ∈ o) := by
+  rw [←Option.mem_toList]
+  infer_instance
```
Comment (eric-wieser) [inline Mathlib/Data/Matrix/PEquiv.lean:56]: This instance shouldn't need `[DecidableEq n]`.

What instance is found in Lean3?

File: Mathlib/Data/Matrix/PEquiv.lean [hunk 3]
```diff
@@ -111,13 +119,15 @@ theorem toMatrix_trans [Fintype m] [DecidableEq m] [DecidableEq n] [Semiring α]
     (g : m ≃. n) : ((f.trans g).toMatrix : Matrix l n α) = f.toMatrix ⬝ g.toMatrix := by
   ext (i j)
   rw [mul_matrix_apply]
-  dsimp [to_matrix, PEquiv.trans]
+  dsimp [toMatrix, PEquiv.trans]
   cases f i <;> simp
 #align pequiv.to_matrix_trans PEquiv.toMatrix_trans
 
+-- Porting note: proof was `rfl`
```
Comment (eric-wieser) [inline Mathlib/Data/Matrix/PEquiv.lean:126]: It would be worth trying to work out why it is no longer `rfl`

Comment (eric-wieser) [reply Mathlib/Data/Matrix/PEquiv.lean:126]: I would guess it's due to the decidable instance being off above.

File: Mathlib/Data/Matrix/PEquiv.lean [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+variable {k l m n : Type _}
+
+variable {α : Type v}
+
+open Matrix
+
+-- Porting note: added
+local instance [DecidableEq n] (j : n) (o : Option n) : Decidable (j ∈ o) :=
+  haveI : Decidable (o = some j) := inferInstance
+  this
```
Comment (eric-wieser) [inline Mathlib/Data/Matrix/PEquiv.lean:56]: This is definitely better, but belongs in the same file that defines the `∈` notation

Comment (Ruben-VandeVelde) [reply Mathlib/Data/Matrix/PEquiv.lean:56]: That's Std.Data.Option.Basic. I'll submit a PR there tomorrow, but wouldn't mind if we moved this along in the meantime.

**Comment by ChrisHughes24** [comment]
bors merge

**Pull request closed without merging**